### ISSUE TITLE
Autocomplete confusion

### ISSUE THREAD

**Issue by zePanzerfaust** [microsoft/pylance-release#152]
VS Code version 1.47.2
Pylance v2020.7.2
"python.languageServer": "Pylance" in global settings.json

```
[Info  - 7:39:51 PM] Pylance language server 2020.7.2 starting
[Info  - 7:39:51 PM] Server root directory: ...
```

I was reading some issues where options had to be changed in order to make autocompletion work correctly. I don't think they apply here, but feel free to correct me :).

Sample project: https://github.com/zePanzerfaust/pylance_autocomplete

scripts/test.py contains calls that raise questions.
Line four works as expected.
Line five is suggested but doesn't work when run with python. (probably because of the import in __init__.py?)
Line seven as expected.
Line nine and ten are not suggested but work when run with python.

The question now is, is this all expected behaviour? I would assume that code that works in python should also be available for autocompletion. In this case it seems packages are never listed. And it would be nice if something doesn't work, that it will not be suggested.

**Comment by erictraut** [comment]
Thanks for reporting the issue.

Python's import system has some very unfortunate design aspects. Namely, an import can have unstated side effects that can affect later imports at runtime. For example, let's say that file1.py contains the statement `import mylib` and later contains the expression `mylib.submodule1`. This will generally result in a runtime exception because `submodule1` is not imported by the top-level `mylib/__init__.py` file. Now, let's assume that file2.py contains the statement `import mylib.submodule1`. If file2.py happens to run before file1.py, the runtime exception in file1.py will disappear because of the side effect of the import in file2.py. No code should rely on these side effects because small changes in initialization ordering or import ordering can expose the problem. Pylance reports these as errors.

Your example did expose two legitimate bugs in Pylance. Both of them were specific to wildcard imports. (Incidentally, I strongly recommend against ever using wildcard imports in user code. They can result in very fragile code.)

1. Pylance was not properly handling the `from .A import *` form within a `__init__.py` file. In this particular case, `A` should be re-exported by the top-level package.

2. Pylance was assuming that in the wildcard case, submodules should be included implicitly as part of the "*". The Python spec doesn't specify, but based on your examples, it's clear that submodules are not implicitly included.

Both of these problems have been addressed, and the fix will be in the next version of Pylance.

**Comment by jakebailey** [comment]
The above fixes didn't quite make it into 2020.7.3, but will appear in the next release.

**Comment by zePanzerfaust** [comment]
Thanks for the quick thorough answer and fix.
